node/core/chain-selection/src/lib.rs

```diff
@@ -0,0 +1,757 @@
+// Copyright 2021 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Implements the Chain Selection Subsystem.
+
+use polkadot_primitives::v1::{BlockNumber, Hash, Header, ConsensusLog};
+use polkadot_subsystem::{
+	Subsystem, SubsystemContext, SubsystemResult, SubsystemError, SpawnedSubsystem,
+	OverseerSignal, FromOverseer,
+	messages::{ChainSelectionMessage, ChainApiMessage},
+	errors::ChainApiError,
+};
+
+use parity_scale_codec::Error as CodecError;
+use futures::channel::oneshot;
+
+use std::collections::HashMap;
+use std::time::{UNIX_EPOCH, SystemTime};
+
+const LOG_TARGET: &str = "parachain::chain-selection";
+
+type Weight = u64;
+type Timestamp = u64;
+
+#[derive(Debug, Clone)]
+enum Approval {
+	// Approved
+	Approved,
+	// Unapproved but not stagnant
+	Unapproved,
+	// Unapproved and stagnant.
+	Stagnant,
+}
+
+impl Approval {
+	fn is_stagnant(&self) -> bool {
+		match *self {
+			Approval::Stagnant => true,
+			_ => false,
+		}
+	}
+}
+
+#[derive(Debug, Clone)]
+struct ViabilityCriteria {
+	// Whether this block has been explicitly reverted by one of its descendants.
+	explicitly_reverted: bool,
+	// `None` means approved. `Some` means unapproved.
+	approval: Approval,
+	earliest_unviable_ancestor: Option<Hash>,
+}
+
+impl ViabilityCriteria {
+	fn is_viable(&self) -> bool {
+		self.is_parent_viable() && self.is_partially_viable()
+	}
+
+	// Whether the current block is partially viable.
+	// That is, whether the current block is neither reverted nor stagnant.
+	fn is_partially_viable(&self) -> bool {
+		!self.explicitly_reverted && !self.approval.is_stagnant()
+	}
+
+	// Whether the parent is viable.
+	fn is_parent_viable(&self) -> bool {
+		self.earliest_unviable_ancestor.is_none()
```

Comment:
The Implementer's Guide mentions the first criteria to be:

`It [the viable block] is or descends from the finalized block`

Do we need to add an additional check to ensure that some parent is the finalized block?

Reply:
I see what you mean. One goal of this implementation is to ensure that only blocks descending from the finalized chain are kept in the DB.